peps/pep-0742.rst

```diff
@@ -0,0 +1,305 @@
+* `Pyright issue <https://github.com/microsoft/pyright/issues/3450>`__ (expecting :py:data:`typing.TypeGuard` to work like :py:func:`isinstance`)
+* `Pyright issue <https://github.com/microsoft/pyright/issues/3466>`__ (expecting narrowing in the ``else`` branch)
+* `Mypy issue <https://github.com/python/mypy/issues/13957>`__ (expecting narrowing in the ``else`` branch)
+* `Mypy issue <https://github.com/python/mypy/issues/14434>`__ (combining multiple TypeGuards)
+* `Mypy issue <https://github.com/python/mypy/issues/15305>`__ (expecting narrowing in the ``else`` branch)
+* `Mypy issue <https://github.com/python/mypy/issues/11907>`__ (user-defined function similar to :py:func:`inspect.isawaitable`)
+* `Typeshed issue <https://github.com/python/typeshed/issues/8009>`__ (``asyncio.iscoroutinefunction``)
+
+Rationale
+=========
+
+The problems with the current behavior of :py:data:`typing.TypeGuard` compel us to improve
+the type system to allow a different type narrowing behavior. :pep:`724` proposed to change
+the behavior of the existing :py:data:`typing.TypeGuard` construct, but we :ref:`believe <pep-742-change-typeguard>`
+that the backwards compatibility implications of that change are too severe. Instead, we propose
+adding a new special form with the desired semantics.
+
+We acknowledge that this leads to an unfortunate situation where there are two constructs with
+a similar purpose and similar semantics. We believe that users are more likely to want the behavior
+of ``TypeNarrower``, the new form proposed in this PEP, and therefore we recommend that documentation
+emphasize ``TypeNarrower`` over ``TypeGuard`` as a more commonly applicable tool. However, the semantics of
+``TypeGuard`` are occasionally useful, and we do not propose to deprecate or remove it. In the long
+run, most users should use ``TypeNarrower``, and ``TypeGuard`` should be reserved for rare cases
+where its behavior is specifically desired.
+
+
+Specification
+=============
+
+A new special form, ``TypeNarrower``, is added to the :py:mod:`typing`
+module. Its usage, behavior, and runtime implementation are similar to
+those of :py:data:`typing.TypeGuard`.
+
+It accepts a single
+argument and can be used as the return type of a function. A function annotated as returning a
+``TypeNarrower`` is called a type narrowing function. Type narrowing functions must return ``bool``
+values, and the type checker should verify that all return paths return
+``bool``. However, ``TypeNarrower`` is not a subtype of ``bool``.
+The type ``Callable[..., TypeNarrower[int]]`` is not assignable to
+``Callable[..., bool]`` or ``Callable[..., TypeGuard[int]]``, and vice versa.
+
+Type narrowing functions must accept at least one positional argument. The type
+narrowing behavior is applied to the first positional argument passed to
+the function. The function may accept additional arguments, but they are
+not affected by type narrowing.
```

Comment:
In PEP 547 there's the phrase

> If a type guard function is implemented as an instance method or class method, the first positional argument maps to the second parameter (after “self” or “cls”).

I'd like TypeNarrower to have the same behavior (unless we've found it unhelpful?) and I think it ought to specified explicitly.

Reply:
I think I dropped that by accident, will re-add it.

Reply:
One criticism that was raised about the original `TypeGuard` formulation is that "first argument" is arguably ambiguous when dealing with class or instance methods. If you invoke an instance method that has been bound to an instance, then the first argument passed to it is associated with the first parameter _after_ `self`. But if you invoke an instance method through its class, then the first argument passed to it is associated with the `self` parameter.

When PEP 647 refers to the "first argument", my intent was that it was the argument that corresponds to the parameter _after_ `self` or `cls` if the type guard function is a method regardless of how this method is invoked. This is how all of the type checkers implemented it as well. Maybe you can think of a clearer way specifying this so as to eliminate all ambiguity.

Reply:
Thanks for the context. I am aiming to make `TypeGuard` and `TypeNarrower` work exactly the same in all aspects other than the type narrowing behavior, because to do otherwise would lead to confusion. I think TypeGuard's wording for self and cls is clear enough, but we can consider improving the wording in the spec at some point.